Update to the latest YCM version + some improvements

drdanz opened the pull request:
* Update YCM files to latest version (0.8.1)
* Use EXPORT instead of TARGETS_PROPERTY in install_basic_package_files
* Require CMake 3.5
* Build position independent code and disable compiler extensions
* Add optional dependency on YCM

drdanz committed:
Update YCM files to latest version

drdanz committed:
Use EXPORT instead of TARGETS_PROPERTY in install_basic_package_files

drdanz committed:
Require CMake 3.5

drdanz committed:
Build position independent code and disable compiler extensions

drdanz committed:
Add comment about using YCM instead of shipping YCM files

traversaro commented:
To be honest, I am not a big fan of the optional dependency on YCM. The relation with YCM and the reason why those files are vendored is clearly explained in the README, and adding the optional dependency on YCM would just make it more difficult to debug the CMake build  script in case some strange version of YCM is installed in the system, without removing the problem that the internal files are  vendored and so they need to manually updated.

drdanz commented:
@traversaro would a comment instead be ok for you?
```cmake
# If you don't want to ship these files with your code (recommended), you can
# instead depend on YCM.
# In this case replace the following line with
# find_package(YCM 0.8.1 REQUIRED)
list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
```

drdanz committed:
Fix RPATH support and remove ENABLE_RPATH option in favour of CMake ones

INSTALL_NAME_DIR should point to the libraries are installed.
Also use the `FULL` version of the GNUInstallDirs variables.

Fixes #7

drdanz commented:
Replaced optional dependency on YCM with a comment.
I also added a new commit, fixing the `RPATH` support (`INSTALL_NAME_DIR` should point to the libraries are installed, i.e. `CMAKE_INSTALL_FULL_LIBDIR`, not to `CMAKE_INSTALL_PREFIX`), and removing the `ENABLE_RPATH` option in favour of CMake standard options (fixes #7).

claudiofantacci commented on CMakeLists.txt:
```diff
@@ -59,12 +55,23 @@ if(MSVC)
 endif()
 
+# Build position independent code
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
```
Since this project is also a tutorial, I'd like to give room to comments to explain why we're setting this feature. For example, we could add something like:

> Position Independent Code (PIC) is commonly used for shared libraries so that the same shared library code can be loaded in each program address space in a location where it will not overlap with any other uses of such memory. In particular, this option avoids problems occurring when a process wants to load more than one shared library at the same virtual address. Since shared libraries cannot predict where other shared libraries could be loaded, this is an unavoidable problem with the traditional shared library concept. Generating position-independent code is often the default behavior for most modern compilers.
Further details on PIC can be found here: https://eli.thegreenplace.net/2011/11/03/position-independent-code-pic-in-shared-libraries/


which is elaborated from Wikipedia. If you have precise hints about this command please feel free to add or modify the description.

claudiofantacci commented on CMakeLists.txt:
```diff
@@ -59,12 +55,23 @@ if(MSVC)
+
+# Disable C and C++ compiler extensions
+set(CMAKE_C_EXTENSIONS OFF)
+set(CMAKE_CXX_EXTENSIONS OFF)
```
Same as above 😄 . I would add something like

> C/CXX_EXTENSIONS are ON by default to allow the compilers to use extended variants of the C/CXX language. However, this could expose cross-platform bugs in user code or in the headers of third-party dependencies and thus it is strongly suggested to turn extensions off.

which is an elaborated version from CMake documentation itself.

claudiofantacci commented on CMakeLists.txt:
```diff
@@ -77,14 +84,10 @@ if(BUILD_TESTING)
+# Enable RPATH support for installed binaries and libraries
+add_install_rpath_support(BIN_DIRS "${CMAKE_INSTALL_FULL_LIBDIR}"
+                          LIB_DIRS "${CMAKE_INSTALL_FULL_BINDIR}"
+                          INSTALL_NAME_DIR "${CMAKE_INSTALL_FULL_LIBDIR}"
```
Awesome, thanks!
I could not be able to find the `CMAKE_INSTALL_FULL_LIBDIR/BINDIR` is it coming as an utility variables from somewhere? If so, could you please add a comment about it specifying what these variables contain.

drdanz replied:
The `CMAKE_INSTALL_FULL_<dir>` variables are well documented in the [GNUInstallDirs documentation](https://cmake.org/cmake/help/git-master/module/GNUInstallDirs.html) I don't think any explaination here is necessary

claudiofantacci replied:
[My tirivial search criteria got me!](https://cmake.org/cmake/help/latest/search.html?q=CMAKE_INSTALL_FULL_LIBDIR&check_keywords=yes&area=default)
Thanks for pointing to GNUInstallDirs documentation 😄

claudiofantacci commented on src/LibTemplateCMake/CMakeLists.txt:
```diff
@@ -2,7 +2,12 @@
+# `install` command), it is necessary to pass the `FIRST_TARGET` option to
+# `install_basic_package_files` in the main CMakeLists.txt file.
+# In this case the project name is used for both, therefore it is not necessary.
+set(LIBRARY_TARGET_NAME ${PROJECT_NAME})
```
Thanks for this nice comment. I'm wondering whether it would be more suitable for a tutorial to explicitly use `FIRST_TARGET` (which I agree would not be necessary here, but is for the sake of teaching 😄) and then write that in this case it would have been sufficient to use the same `PROJECT_NAME` without `FIRST_TARGET`.
Either way, I'm fine so it is up to you to choose 😄

drdanz replied:
I considered that as well, but I think in this case it is not really useful, it does not teach you anything more than what is explained in the comment. Also if you need to do something more complicated, you can easily read the documentation of the modules...

claudiofantacci reviewed (changes requested):
Thanks for the contribution!
I just wrote down a copule of comments of which I would like to know your opinion.

claudiofantacci commented:
Please also note that @mbrunettini is doing some tests with Genkins on master branch, so please do not merge this until he finishes.

Please @mbrunettini let us know when you are ok with the test so that we can clean master commits and merge this PR!

Thanks everyone for the collaboration 🎉

drdanz committed:
Extend comments and add links to YCM modules

drdanz commented:
Updated with the comments requested (I added an extra comment on the PIC one about shared libraries linking non-PIC static ones) and added links to YCM modules documentation

claudiofantacci reviewed (dismissed):
Great! Thanks for changes and the discussion.

claudiofantacci commented:
Please don't merge until @mbrunettini has finished with Jenkins tests.

traversaro commented on CMakeLists.txt:
```diff
@@ -77,14 +103,11 @@ if(BUILD_TESTING)
-                          LIB_DIRS "${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}"
-                          INSTALL_NAME_DIR "${CMAKE_INSTALL_PREFIX}"
-                          DEPENDS ENABLE_RPATH
+add_install_rpath_support(BIN_DIRS "${CMAKE_INSTALL_FULL_LIBDIR}"
```
What's the option to enable/disable install rpath equivalent to `ENABLE_RPATH` ? Even if it is a CMake variable, should we explicitly list as an option as we do for `BUILD_SHARED_LIBS` ?

claudiofantacci replied:
It is `CMAKE_SKIP_RPATH` which automatically pops out at configuration time.
If we make it as an option it would probably be redundand, isn't it @drdanz ?

claudiofantacci commented:
@drdanz can you please rebase on master and remove the commits related to the Jenkins files? After that I merge the PR. Thanks!

drdanz commented:
@claudiofantacci done.

claudiofantacci commented:
Thanks @drdanz!
